### PR TITLE
Support GHC 9.2

### PR DESCRIPTION
**Pull request by RyanGlScott**
This contains a variety of minor changes needed to make `saw` and its supporting libraries compile with GHC 9.2:

* One place in `heapster-saw` matches on a GADT using a `let` binding, which no longer typechecks in GHC 9.2. While somewhat mysterious, using `let` bindings to match on GADTs is fragile anyway, so I opted to simply replace this with a monadic `do` match, which does not exhibit the same fragility.
* GHC now includes `-Wincomplete-uni-patterns` in `-Wall` as of 9.2, so much of this patch is dedicated to silencing these sorts of warnings, usually by adding explicit fall-through cases for partial pattern matches.
* GHC now inclues `-Wnoncanonical-monad-instances` in `-Wall` as of 9.2, so I had to refactor some `Applicative`/`Monad` instances to fix these warnings.
* GHC 9.2 now defines `readBin` in `Numeric`, which clashes with a function of the same name in `SAWScript.Lexer`. Using explicit imports avoids this.
* GHC's pattern-match coverage checker is smarter in 9.2 (see [here](https://gitlab.haskell.org/ghc/ghc/-/wikis/migration/9.2?version_id=63085dd8a5b56370571bda428848e7098765f7f8#improved-pattern-match-coverage-checker)), so I had to use `EmptyCase` to silence some of the new `-Wincomplete-patterns` warnings that were uncovered in `heapster-saw`.

### CODE REVIEW

**Comment by RyanGlScott** [comment]
At long last, I believe this is finally ready. A special thanks to @eddywestbrook for contributing patches to fix the numerous `-Wincomplete-uni-patterns` warnings in `heapster-saw` that were uncovered by GHC 9.2 adding that warning to `-Wall`. I also tried to clean up the uses of `head` and `tail` to replace them with more specific versions of these functions that give slightly more specific errors if they fail—let me know if you are OK with this.

I've also added @chameco and @kquick as reviewers for everything else in this patch. It touches a lot of different parts of the code, but it's mostly a collection of small changes. Let me know if anything looks especially surprising!

**Review by eddywestbrook — APPROVED** [review]
After going through these changes in a relatively high level of detail, the changes overall look good and make sense, though I haven't been able to look at absolutely everything. At a high level, the changes mostly are to address the new requirement that pattern-matching always satisfies coverage, and raise explicit errors when an unexpected pattern is matched.

File: heapster-saw/src/Verifier/SAW/Heapster/Permissions.hs [hunk 1]
```diff
@@ -3115,13 +3115,16 @@ addPermOffsets (LLVMPermOffset off1) (LLVMPermOffset off2) =
 
 -- | Get the @n@th expression in a 'PermExprs' list
 nthPermExpr :: PermExprs args -> Member args a -> PermExpr a
+nthPermExpr PExprs_Nil m = case m of {}
```
Comment (kquick) [inline heapster-saw/src/Verifier/SAW/Heapster/Permissions.hs:3118]: Same as above: should this be `error` or `panic`?

File: heapster-saw/src/Verifier/SAW/Heapster/Permissions.hs [hunk 2]
```diff
@@ -3115,13 +3115,16 @@ addPermOffsets (LLVMPermOffset off1) (LLVMPermOffset off2) =
...
 nthPermExpr (PExprs_Cons _ arg) Member_Base = arg
 nthPermExpr (PExprs_Cons args _) (Member_Step memb) =
   nthPermExpr args memb
 
 -- | Set the @n@th expression in a 'PermExprs' list
 setNthPermExpr :: PermExprs args -> Member args a -> PermExpr a ->
                   PermExprs args
+setNthPermExpr PExprs_Nil m _ =
+  case m of {}
```
Comment (kquick) [inline heapster-saw/src/Verifier/SAW/Heapster/Permissions.hs:3127]: Same comment.

File: heapster-saw/src/Verifier/SAW/Heapster/Permissions.hs [hunk 3]
```diff
@@ -7204,6 +7207,7 @@ permVarSubstToNames PermVarSubst_Nil = MNil
 permVarSubstToNames (PermVarSubst_Cons s n) = permVarSubstToNames s :>: n
 
 varSubstLookup :: PermVarSubst ctx -> Member ctx a -> ExprVar a
+varSubstLookup PermVarSubst_Nil m = case m of {}
```
Comment (kquick) [inline heapster-saw/src/Verifier/SAW/Heapster/Permissions.hs:7210]: Same comment.

File: src/SAWScript/Interpreter.hs [hunk 4]
```diff
@@ -19,7 +19,8 @@ Stability   : provisional
 {-# LANGUAGE ScopedTypeVariables #-}
 {-# LANGUAGE TupleSections #-}
 {-# LANGUAGE ViewPatterns #-}
-{-# LANGUAGE NondecreasingIndentation #-}
+-- See Note [-Wincomplete-uni-patterns and irrefutable patterns] in SAWScript.MGU
+{-# OPTIONS_GHC -Wno-incomplete-uni-patterns #-}{-# LANGUAGE NondecreasingIndentation #-}
```
Comment (kquick) [inline src/SAWScript/Interpreter.hs:23]: formatting

**Review by kquick — APPROVED** [review]
All the trailing whitespace and corresponding removals are unfortunate, but for the most part the changes here are pretty isomorphic to the original implicit behavior, so I don't see anything concerning.

**Pull request merged**